File: UseCases/src/codes/seanhenry/mockgenerator/util/UniqueMethodNameGenerator.java
package codes.seanhenry.mockgenerator.util;

import java.util.*;
import java.util.stream.Stream;

public class UniqueMethodNameGenerator {

  private final HashMap<String, String> uniqueMethodName = new HashMap<>();
  private final Set<MethodModel> duplicateMethodModels = new HashSet<>();

  public UniqueMethodNameGenerator(MethodModel... methodModels) {
    this(Arrays.asList(methodModels));
  }

  public UniqueMethodNameGenerator(List<MethodModel> methodModels) {
    duplicateMethodModels.addAll(methodModels);
  }

  public void generateMethodNames() {
    while (!duplicateMethodModels.isEmpty()) {
      processDuplicates();
    }
  }

  private void processDuplicates() {
    HashMap<String, List<MethodModel>> nameBuckets = moveDuplicatesToNameBuckets();
    for (String name : new HashSet<>(nameBuckets.keySet())) {
      List<MethodModel> models = nameBuckets.get(name);
      stripUniqueModels(name, models);
    }
  }

  private void stripUniqueModels(String name, List<MethodModel> models) {
    if (models.size() == 1) {
      confirmModelIsUnique(name, models.get(0));
    } else {
      sortBySimplest(models);
      MethodModel simplestModel = models.get(0);
      if (isUniquelySimple(models)) {
        confirmModelIsUnique(name, simplestModel);
      }
      streamMaximumComplexityModels(models)
        .forEach(m -> confirmModelIsUnique(name, m));
    }
  }

  private static Stream<MethodModel> streamMaximumComplexityModels(List<MethodModel> models) {
    return models
      .stream()
      .filter(UniqueMethodNameGenerator::canModelGetMoreComplex);
  }

  private static boolean canModelGetMoreComplex(MethodModel model) {
    return model.peekNextPreferredName() == null;
  }

  private static boolean isUniquelySimple(List<MethodModel> models) {
    int simplestParamCount = models.get(0).getParameterCount();
    int nextSimplestParamCount = models.get(1).getParameterCount();
    return simplestParamCount < nextSimplestParamCount;
  }

  private void confirmModelIsUnique(String name, MethodModel simplestModel) {
    duplicateMethodModels.remove(simplestModel);
    uniqueMethodName.put(simplestModel.getID(), name);
  }

  private static void sortBySimplest(List<MethodModel> models) {
    models.sort(Comparator.comparingInt(MethodModel::getParameterCount));
  }

  private HashMap<String, List<MethodModel>> moveDuplicatesToNameBuckets() {
    HashMap<String, List<MethodModel>> nameBuckets = new HashMap<>();
    for (MethodModel model : duplicateMethodModels) {
      String name = model.nextPreferredName();
      List<MethodModel> models = getModels(nameBuckets, name);
      models.add(model);
      nameBuckets.put(name, models);
    }
    return nameBuckets;
  }

  private static List<MethodModel> getModels(HashMap<String, List<MethodModel>> nameBuckets, String name) {
    List<MethodModel> models = nameBuckets.get(name);
    if (models == null) {
      models = new ArrayList<>();
    }
    return models;
  }

  public String getMethodName(String id) {
    return uniqueMethodName.get(id);
  }
}

File: src/main/kotlin/codes/seanhenry/mockgenerator/util/UniqueMethodNameGenerator.kt
package codes.seanhenry.mockgenerator.util

class UniqueMethodNameGenerator(methodModels: List<MethodModel>) {

  private val uniqueMethodName = HashMap<String, String>()
  private val duplicateMethodModels = HashSet<MethodModel>()

  constructor(vararg methodModels: MethodModel) : this(listOf(*methodModels))

  init {
    duplicateMethodModels.addAll(methodModels)
  }

  fun generateMethodNames() {
    while (!duplicateMethodModels.isEmpty()) {
      processDuplicates()
    }
  }

  private fun processDuplicates() {
    val nameBuckets = moveDuplicatesToNameBuckets()
    nameBuckets.forEach {
      commitUniqueModels(it.key, it.value)
    }
  }

  private fun commitUniqueModels(name: String, models: MutableList<MethodModel>) {
    if (models.size == 1) {
      commitModel(name, models[0])
    } else {
      sortBySimplest(models)
      val simplestModel = models[0]
      if (isUniquelySimple(models)) {
        commitModel(name, simplestModel)
      }
      commitModelsThatCannotGetMoreComplex(models, name)
    }
  }

  private fun commitModelsThatCannotGetMoreComplex(models: List<MethodModel>, name: String) {
    models
        .filter { canModelGetMoreComplex(it) }
        .forEach { commitModel(name, it) }
  }

  private fun canModelGetMoreComplex(model: MethodModel): Boolean {
    return model.peekNextPreferredName() == null
  }

  private fun isUniquelySimple(models: List<MethodModel>): Boolean {
    val simplestParamCount = models[0].parameterCount
    val nextSimplestParamCount = models[1].parameterCount
    return simplestParamCount < nextSimplestParamCount
  }

  private fun commitModel(name: String, simplestModel: MethodModel) {
    duplicateMethodModels.remove(simplestModel)
    uniqueMethodName[simplestModel.id] = strip(name)
  }

  private fun strip(name: String): String {
    return name.replace("\\W".toRegex(), "")
  }

  // sortBy not available in Kotlin Native

  private fun sortBySimplest(models: MutableList<MethodModel>) {
    var swapped: Boolean
    do {
      swapped = false
      for (i in 1 until models.size) {
        if (models[i].parameterCount < models[i - 1].parameterCount) {
          val m = models[i - 1]
          models[i - 1] = models[i]
          models[i] = m
          swapped = true
        }
      }
    } while (swapped)
  }

  private fun moveDuplicatesToNameBuckets(): HashMap<String, MutableList<MethodModel>> {
    val nameBuckets = HashMap<String, MutableList<MethodModel>>()
    for (model in duplicateMethodModels) {
      val name = model.nextPreferredName()!!
      val models = getModels(nameBuckets, name)
      models.add(model)
      nameBuckets[name] = models
    }
    return nameBuckets
  }

  private fun getModels(nameBuckets: HashMap<String, MutableList<MethodModel>>, name: String?): MutableList<MethodModel> {
    var models: MutableList<MethodModel>? = nameBuckets[name]
    if (models == null) {
      models = ArrayList()
    }
    return models
  }

  fun getMethodName(id: String): String? {
    return uniqueMethodName[id]
  }
}

